chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,275 @@
+            aug_msg.extend_from_slice(msg.borrow()); // pk + msg
+            let mut hasher = Sha256::new();
+            hasher.update(aug_msg);
+            let h: Bytes32 = hasher.finalize().into();
+            let pairing: Option<&GTElement> = self.cache.get(&h);
+            match pairing {
+                Some(pairing) => {
+                    if !force_cache {
+                        // Heuristic to avoid more expensive sig validation with pairing
+                        // cache when it's empty and cached pairings won't be useful later
+                        // (e.g. while syncing)
+                        missing_count += 1;
+                        if missing_count > pks.borrow().len() / 2 {
+                            return vec![];
+                        }
+                    }
+                    pairings.push(Some(pairing.clone()));
+                },
+                _ => {
+                    pairings.push(None);
+                },
+            }
+            
+        }
+
+        // G1Element.from_bytes can be expensive due to subgroup check, so we avoid recomputing it with this cache
+        let mut pk_bytes_to_g1: HashMap<Bytes48, PublicKey> = HashMap::new();
+        let mut ret: Vec<GTElement> = vec![];
+
+        for (i, pairing) in pairings.iter_mut().enumerate() {
+            if let Some(pairing) = pairing {  // equivalent to `if pairing is not None`
+                ret.push(pairing.clone());
+            } else {
+                let mut aug_msg = pks.borrow()[i].to_vec();
+                aug_msg.extend_from_slice(&msgs.borrow()[i]);  // pk + msg
+                let aug_hash: Signature = hash_to_g2(&aug_msg);
+
+                let pk_parsed: &mut PublicKey = pk_bytes_to_g1.entry(pks.borrow()[i]).or_insert_with(|| {
+                    PublicKey::from_bytes(&pks.borrow()[i]).unwrap()
+                });
+
+                let pairing: GTElement = aug_hash.pair(pk_parsed);
+                let mut hasher = Sha256::new();
+                hasher.update(&aug_msg);
+                let h: Bytes32 = hasher.finalize().into();
+                self.cache.put(h, pairing.clone());
+                ret.push(pairing);
+            }
+        }
+
+        ret
+    }
+
+    pub fn aggregate_verify(
+        &mut self,
+        pks: &Vec<Bytes48>,
+        msgs: &Vec<Vec<u8>>,
+        sig: &Signature,
+        force_cache: bool, 
+    ) -> bool {
+        let mut pairings: Vec<GTElement> = self.get_pairings(pks, msgs, force_cache);
+        if pairings.is_empty() {
+            let mut data = Vec::<(PublicKey, Vec<u8>)>::new();
+            for (pk, msg) in pks.iter().zip(msgs.iter()) {
+                let pk = PublicKey::from_bytes_unchecked(pk).unwrap();
+                data.push((pk.clone(), msg.clone()));
+            }
+            let res: bool = agg_ver(sig, data);
+            return res
+        }
+        let pairings_prod = pairings.pop(); // start with the first pairing
```

Comment:
Note that the `GT` group has an "identity" element which acts like `1` for multiplication of "regular" (ie. real) numbers. If you set the partial product `prod` to this identity element, you don't have to do this weird `.pop` which treats the first argument as special. In fact, you can probably use `.fold` https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold

There are still two cases: an empty list and a non-empty list. It's a bit weird to me that you don't check the signature for the empty list case, but it might be a consensus change to do so depending on how empty blocks are treated (ie. blocks that don't have spends with any `AGG_*` conditions).